Intro to lit updates

arthurevans opened the pull request:
Fixes several reported issues and a couple of unreported issues in the intro tutorial. 

- Removes `requestUpdate()` call and favors immutable data pattern instead.
- Adds missing export statements and other inconsistencies noted in #703.
- Fixes a confusingly-written note reported in #753.
- Fixes two errors in JS starting files that I noticed while making the other changes.

arthurevans commented on packages/lit-dev-content/samples/tutorials/intro-to-lit/01/before/my-element.js:
```diff
@@ -1 +1,2 @@
-export {};
+import {LitElement, html} from 'lit';
+import {customElement} from 'lit/decorators.js';
```
```suggestion
import {LitElement, html} from 'lit';
```

Oops. Might've been in a bit of a hurry there.

arthurevans commented on packages/lit-dev-content/site/tutorials/content/intro-to-lit/03.md:
````diff
@@ -8,13 +8,13 @@ Here we've provided a name tag component with a message and an input element. In
 
-    ```html
-    @input=${this.changeName}
+    ```js
````
```suggestion
    ```html
```

e111077 commented on packages/lit-dev-content/site/tutorials/content/intro-to-lit/01.md:
```diff
@@ -19,7 +19,7 @@ In Lit, most things start with defining a _component_. Here we've given you a st
     {% endswitchable-sample %}
 
-    The `MyElement` class provides the implementation for your new component, and the `@customElement` decorator registers it with the browser as a new element type named `my-element`.
+    The `MyElement` class provides the implementation for your new component. <ts-js><span slot="ts">The <code>@customElement</code> decorator registers it with the browser as a new element type named <code>my-element</code>.</span><span slot="js">The <code>customElements.define</code> call registers it with the browser as a new element type named <code>my-element</code>.</span></ts-js>
```
The HTML minifier unfortunately deletes this space in the prod build (it doesn't run in the dev build). We just need to add a `&nbsp;` to force a space there, minification or not.

```suggestion
    The `MyElement` class provides the implementation for your new component.&nbsp;<ts-js><span slot="ts">The <code>@customElement</code> decorator registers it with the browser as a new element type named <code>my-element</code>.</span><span slot="js">The <code>customElements.define</code> call registers it with the browser as a new element type named <code>my-element</code>.</span></ts-js>
```

e111077 commented on packages/lit-dev-content/site/tutorials/content/intro-to-lit/01.md:
```diff
@@ -36,3 +36,12 @@ In Lit, most things start with defining a _component_. Here we've given you a st
+
+The TypeScript version of this code uses the `@customElement` decorator to register your class with the browser as a new element. The JavaScript version uses `customElements.define()` to do the same thing. This is a matter of style. You can use decorators—or not use decorators—in either language. For more information, see [Decorators](/docs/components/decorators/).
+
+{% endaside %}
```
is it worth mentioning the TS/JS toggle. Surprising amount of people online don't notice that it exists and have issues with some of our decorators. Specifically `@query`

e111077 replied:
ah i see you talk about query a later step 👍 👍

arthurevans replied:
Yeah, I talk about that but mentioning the toggle explicitly is a good idea. It was added after the tutorial was written, so there were a number of confusing bits.

arthurevans replied:
Added note, with a little extra switchable sample as a demo.

e111077 commented on packages/lit-dev-content/site/tutorials/content/intro-to-lit/03.md:
````diff
@@ -8,13 +8,13 @@ Here we've provided a name tag component with a message and an input element. In
     ```
 
-    <code>@<var>eventName</var></code> is a special syntax for adding an event handler using an expression.
+    This `@input` expression adds `this.changeName` as an event listener for the `input` event. (You can use this syntax for any event listener—just replace input with any event name and `this.changeName` with any JavaScript expression that evaluates to an event handler function.)
````
`event handler function` -> `event handler`. Because AFAIK the only way to do 

`el.addEventListener('name', 'props', {passive: true});`

in javascript is to use an event handler object:

```
const clickHandler = {

    // This fires when the handler is called
    handleEvent(e) { console.log('clicked'); }

    // You can specify event options same as addEventListener
    capture: false;
    passive: true;
}

html`<button @click=${clickHandler}>Click Me</button>`
```

https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#the_event_listener_callback

arthurevans replied:
👍

arthurevans replied:
```suggestion
    This `@input` expression adds `this.changeName` as an event listener for the `input` event. (You can use this syntax for any event listener—just replace input with any event name and `this.changeName` with any JavaScript expression that evaluates to an event handler.)
```

e111077 commented on packages/lit-dev-content/site/tutorials/content/intro-to-lit/05.md:
```diff
@@ -38,21 +38,32 @@ You can use standard JavaScript in Lit expressions to create conditional or repe
 
     addToDo() {
-      this.listItems.push({text: this.input.value, completed: false});
+      this.listItems = [...this.listItems,
+          {text: this.input.value, completed: false}];
```
I still personally disagree with this method being the one we choose to highlight.

Especially since it's an input `@property` to the component – any references the user of this component hands to this component is no longer the one used inside this component. The user then has to query the component again and get the new refernce. Then there's idea of cloning objects etc.

arthurevans replied:
Yeah, I can see the argument here with regard to it being an input property. It would probably make more sense for it to *not* be an input property, since the component manipulates it. 

@kevinpschaaf and @sorvell both felt we should not be recommending `requestUpdate()` for most use cases.

e111077 replied:
This is more of a "alright but I'mma complain the entire time" feel free to ignore the noise 🌬️

sorvell replied:
I think we should make this `@state` and then split this into 2 steps. One just intro'ing the `@state` and adding the list and the other doing the `add` stuff.

kevinpschaaf replied:
If it's too big of a change to split out another step, an aside that says something like this seems fine:

> Notice that `listItems` uses the `@state()` decorator. That makes it reactive just like `@property()`, but is internal to our element, not part of the public API of our element. Read more about [Internal reactive state](https://lit.dev/docs/components/properties/#internal-reactive-state).

arthurevans replied:
I think this latter should work. It'll just be a matter of making sure I fix all the references to listItems. Which should probably also become private or protected `_listItems`?

arthurevans replied:
Made this change, and added a new paragraph about state. The asides just seem way to distracting to put inline.

e111077 reviewed (commented):
Leaving these as comment so I don't block the merge since i'm OOO on mon-tues

sorvell commented on packages/lit-dev-content/samples/tutorials/intro-to-lit/05/after/todo-list.ts:
```diff
@@ -2,12 +2,12 @@ import {LitElement, html} from 'lit';
-class ToDoList extends LitElement {
-  @property()
+export class ToDoList extends LitElement {
+  @property({attribute: false})
```
This should be `@state`

sorvell commented on packages/lit-dev-content/site/tutorials/content/intro-to-lit/05.md:
````diff
@@ -38,21 +38,32 @@ You can use standard JavaScript in Lit expressions to create conditional or repe
     ```js
-    this.input = this.shadowRoot.querySelector('#newitem');
+    get input() {
+      return this.renderRoot?.querySelector('#newitem') ?? null;
````
Why change this to use `renderRoot`?

arthurevans replied:
I'm just adding the code that our TS->JS translator generates so this code matches the solved code.

arthurevans commented:
Made the requested changes:

- Changed `listItems` to internal state, named `_listItems`.
- Added a note about the JS/TS language switch on the intro page.

Also updated the last page to link to new tutorials.

arthurevans commented:
This is ready to go as soon as #794 merges.

e111077 commented:
Waiting for https://github.com/lit/lit.dev/pull/820#pullrequestreview-966868027

Are we doing this? 👏👏👏

arthurevans commented:
I added some links to the new tutorials, so I think we should do these basically at the same time. Or this one can even follow #794, there's no hard requirement.
